[MRG] add sqlite3 implementations for `Index`, `CollectionManifest`, and `LCA_Database`

ctb opened the pull request:
This PR adds three new SQLite-based features to sourmash, `SqliteIndex`, `SqliteCollectionManifest`, and `LCA_SqliteDatabase`. These new features provide the following features (in order of their estimated importance to the future world domination of sourmash):

* fast SQLite-based manifest support for extremely large collections
* low-memory, low-latency, and multiprocess LCA databases for better LCAing,
* a more unified approach to SQLite database handling in sourmash,
* a reasonably nice alternative to other indexed collections;

Fixes https://github.com/sourmash-bio/sourmash/issues/1807
Fixes https://github.com/sourmash-bio/sourmash/issues/1930
Fixes https://github.com/sourmash-bio/sourmash/issues/948 by providing a faster `LCA_Database`
Fixes https://github.com/sourmash-bio/sourmash/issues/821
Fixes https://github.com/sourmash-bio/sourmash/issues/1111 by using SQLite rather than JSON.

Provides one solution for https://github.com/sourmash-bio/sourmash/issues/909, a shared LCA index for many processes.

## Highlights

**`SqliteIndex`** is a SQLite based Index class that:
* stores sketches and hashes in a sqlite db;
* is restricted to storing scaled sketches with a single scaled, albeit with multiple ksizes and moltypes;
* does efficient, on-the-fly downsampling of sketches;

**`SqliteCollectionManifest`** is a SQLite based Manifest class that:
* stores manifest entries in sqlite;
* handles selection and picklists appropriately, in SQL (so, fast!);
* joins `CollectionManifest` in being loadable via the CLI in a `StandaloneManifestIndex` as a first-class Index.

**`LCA_SqliteDatabase`** is a SQLite-based LCA class built on top of `SqliteIndex` and `LineageDB_Sqlite` that supports all of the LCA commands.

This PR introduces much of the remaining functionality from "manifests of manifests" https://github.com/sourmash-bio/sourmash/issues/1652 https://github.com/sourmash-bio/sourmash/issues/1685 into sourmash, after merge of `StandaloneManifestIndex` in #1891.

## Reading and writing

All classes are fully integrated into the sourmash command-line functionality.

Collections of all three types can be read as regular ol' `Index` classes using `sourmash.load_file_as_index(...)`.

SQLite databases can be output with `-o <name>.sqldb`, and are read natively on the command line.

SQLite manifests can be output by passing `-F sql` to `sourmash sig manifest`, and will be read natively by sourmash as `StandaloneManifestIndex`.

SQLite LCA databases can be output by passing `-F sql` to `sourmash lca index`.

## Other additions

This PR also:
* fixes GitHub workflow caching to track setup.cfg, and provides a version number to manually increment when new dependencies are added;
* adds the `-d/--debug` flag to `sourmash search`
* fixes a bug where `LCA_Database.signatures(...)` didn't pay attention to picklists.

## Implementation notes

`SqliteIndex` is built on `SqliteCollectionManifest`, and `SqliteCollectionManifest` is usable without `SqliteIndex`.

`SqliteIndex` and `SqliteCollectionManifest` share  the `sketches` table.

`SqliteIndex` supports a scaled of 1 - that is, direct storage of unsigned 64-bit numbers! - via the `bitstring` library, which converts unsigned long longs into signed long longs that sqlite can store. **This is a new dependency.**

## SQLite databases, and backwards compatibility

This PR also adds a required table to all sourmash SQLite databases: the table `sourmash_internal` contains key, value pairs that document what version of `SqliteIndex`, `SqliteManifest`, and `SqliteLineage` (taxonomy) are supported by the database. See the module `sourmash.sqlite_utils` for supporting code.

The only previous use of SQLite in sourmash was in taxonomy databases, for the `sourmash.tax` module, output by `sourmash tax prepare`. This code used the table `taxonomy`. Moving forward this table has been renamed to `sourmash_taxonomy` and taxonomy SQLite databases also contain the `sourmash_internal` table with a `SqliteLineage` entry. The previous use is fully supported by backwards compatibility code.

## Technical notes

(the below is also enshrined in the docstring in `index/sqlite_index.py`) -

```
sqlite3 based Index, CollectionManifest, and LCA_Database
implementations.

These classes support a variety of flexible and fast on-disk storage,
search, and retrieval functions.

SqliteIndex stores full scaled signatures; sketches are stored as
reverse-indexed collections of hashes. Search is optimized via the
reverse index. Num and abund sketches are not supported. All scaled
values must be the same upon insertion. Multiple moltypes _are_
supported.

SqliteCollectionManifest provides a full implementation of the
manifest API. It can store details for all signature types. When used
as part of a SqliteIndex database, it does not support independent
insertion.

LCA_SqliteDatabase builds on top of SqliteIndex and LineageDB_Sqlite
(in the tax submodule) to provide a full on-disk implementation of
LCA_Database.

Using these classes
-------------------

These classes are fully integrated into sourmash loading.
Internally, use `sqlite_index.load_sqlite_index(...)` to load a specific
file; this will return the appropriate SqliteIndex, StandaloneManifestIndex,
or LCA_Database object.
Use `CollectionManifest.load_from_filename(...)` to load the manifest
directly as a manifest object.

Implementation Details
----------------------

SqliteIndex:
* Hashes with values above MAX_SQLITE_INT=2**63-1 are transformed into
  signed long longs upon insertion, and then back into ulong longs upon
  retrieval.
* Hash overlap is calculated via a SELECT.
* SqliteIndex relies on SqliteCollectionManifest for manifest functionality,
  including signature selection and picklists.
SqliteCollectionManifest:
* each object maintains info about whether it is being "managed" by a
  SqliteIndex class or not. If it is, `_insert_row(...)` cannot be
  called directly.
* `select(...)` operates directly with SQL queries, except for
  picklist selection, which involves inspect each manifest row in
  Python. In addition to being (much) simpler, this ends up being
  faster in some important real world situations, even for millions of
  rows!
* filter_on_rows and filter_on_columns also both operate in Python,
  not SQL.
* for this reason, the `locations()` method returns a superset of
  locations.  This is potentially very significant if you do a select
  with a picklist that ignores most sketches - the `locations()`
  method will ignore the picklist.
  
Limitations:
* all of these classes share a single connection object, and it could
  get confusing quickly if you simultaneously insert and query. We suggest
  separating creation and insertion. That having been said, these databases
  should work fine for many simultaneous queries; just don't write :).
```
TODO testing: test internal and command line for,
- [x] creating an index
- [x] loading an index
- [x] loading an index as a manifest (should work)
- [x] loading an index as a standalone manifest and inserting (should fail)
- [x] creating a manifest, using from CLI
- [x] loading a manifest as a standalone index
- [x] loading a manifest as a standalone index in wrong directory
- [x] loading a manifest as a standalone index and insert (should succeed)
- [x] loading/using a checked-in index
- [x] loading/using a checked-in manifest
- [x] loading a lineage db with old table
- [x] loading a lineage db with new table name/sourmash info
- [x] loading a checked-in lineage db with old table
- [x] loading a checked-in lineage db with new table
- [x] do some realistic benchmarking of SqliteIndex and LCA_Database
- [x] check LCA database is loaded by load_sqlite_index.
- [x] check 'summarize' output on all three
- [x] test identifiers with '.' in sql lca dbs...
- [X] test LCA_SqliteDatabase.open with a SqliteIndex.
- [X] test LCA_SqliteDatabase.open with an empty SqliteIndex
- [x] implement update lca DB/on disk insert stuff for lca index?
- [x] test LCA_SqliteDatabase with a different lineagedb class?
- [X] add `-F sql` to sig check for manifest output
- [x] test `SaveSignatures` SQLite class repr.
- [x] test `db_outfile += '.lca.json'` line in lca_db.py
- [x] test `raise Exception(f"unknown save format for LCA_Database: '{format}'")` in lca_db.py
- [x] test `append= True` in manifest.py
- [x] test `__eq__` in CollectionManifest where manifests are NOT equal`p
- [x] test sqlite_utils/bad version code
- [x] test tax_utils/bad version code
- [x] test tax_utils/no version code
- [x] Write more about new `sourmash_internal` SQLite table, and also how the various sqlite things are compatible.
- [x] document new taxonomy foo in PR description

ctb committed:
switch to get_matching_sketches

ctb committed:
change default cache size

ctb committed:
count overlaps in SQL?

ctb committed:
initial addition of 'sig fileinfo'

ctb committed:
finish first-draft implementation of fileinfo and get_manifest

ctb committed:
cleanup and move over to sourmash_args

ctb committed:
add manifest and length support to LCA_Database

ctb committed:
add rebuild/no-rebuild args

ctb committed:
Merge branch 'latest' of https://github.com/dib-lab/sourmash into add/sqlite_index

ctb committed:
Merge branch 'add/sig_fileinfo' into add/sqlite_index_bitflip

ctb committed:
use BitArray to convert uint to int

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
cleanup

ctb committed:
fix the things?

ctb committed:
Merge branch 'add/sqlite_index_bitflip2' into add/sqlite_index

ctb committed:
cleanup

ctb committed:
more cleanup

ctb committed:
flag when scores are diff

ctb committed:
fix __len__ for zipfiles, __bool__ interpretation

ctb committed:
Merge branch 'add/sig_fileinfo' into add/sqlite_index

ctb committed:
add more index, etc

ctb committed:
more cleanup

ctb committed:
correct for rust panic a la zip

ctb committed:
commit every so often...

ctb committed:
add some comments

ctb committed:
Merge branch 'add/sig_fileinfo' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
get basic manifest-generating machinery working

ctb committed:
update manifest stuff

ctb committed:
add bitstring in support of SqliteIndex

ctb committed:
more cleanup

ctb committed:
add more tests

ctb committed:
add conditions to _get_matching_sketches

ctb committed:
remove conditions

ctb committed:
Merge branch 'add/sqlite_index' of https://github.com/dib-lab/sourmash into add/sqlite_index

ctb committed:
Merge branch 'add/sqlite_index' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
remove errant raise

ctb committed:
update structure

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
some commentary

ctb committed:
switch over to debug_literal

ctb committed:
switch to debug_literal; test tricky ordering

ctb committed:
add LCA database test for tricky ordering

ctb committed:
add LCA database test for tricky ordering

ctb committed:
add test for jaccard ordering to SBTs

ctb committed:
add test for jaccard ordering to SBTs

ctb committed:
Merge branch 'add/test_jaccard_ordering' into add/sqlite_index

ctb committed:
add bitstring to setup

ctb committed:
factor out CollectionManifest_Sqlite

ctb committed:
some basic manifests

ctb committed:
add sqlite manifest rows interface

ctb committed:
minor refactor

ctb committed:
support sig manifest / test it

ctb committed:
move row insert into manifest class

ctb committed:
test creation of sqlite mf

ctb committed:
switch to explicit moltype

ctb committed:
cleanup and refactoring

ctb committed:
cleanup

ctb committed:
SQLite manifests are now first class

ctb committed:
pip cache should be looking at setup.cfg I think?

ctb committed:
and tox cache should be looking at setup.cfg, too

ctb committed:
try again/invalidate cache

ctb committed:
try again

ctb committed:
remove print

ctb committed:
fix some stuff

ctb committed:
even more

ctb committed:
add 'sourmash_versions' table

ctb committed:
test direct sqlmf creation & loading

ctb committed:
improve version checkingc

ctb committed:
test various insertion errors

ctb committed:
fix num support in sqlite manifests (but not index)

ctb committed:
add explicit validation code, to be removed later

ctb committed:
explicit check of 'num'

ctb committed:
add more docs/notes/annotations for work

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
rename CollectionManifest_Sqlite to SqliteCollectionManifest

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
preliminary victory over rankinfo

ctb committed:
provide generic LCA Database functionality via sqlite

ctb committed:
refactor and comment

ctb committed:
refactor and document

ctb committed:
add sqlite_utils

ctb committed:
cleanup

ctb committed:
parse out SqliteIndex.create

ctb committed:
rm comment

ctb committed:
add database_format to lca index

ctb committed:
get sql database output working for LCA index

ctb committed:
get all lca tests working on SQL version of LCA_Database

ctb committed:
add test_index_protocol

ctb committed:
add tests of indices after save/load

ctb committed:
match Index definition of __len__ in sbt

ctb committed:
Merge branch 'add/test_jaccard_ordering' into add/index_tests

ctb committed:
more index tests

ctb committed:
add some generic manifest tests

ctb committed:
define abstract base class for CollectionManifest

ctb committed:
fix GTDB example, sigh

ctb committed:
test hashval_to_idx

ctb committed:
add actual test for min num in rankinfo

ctb committed:
update 'get_lineage_assignments' in lca_db

ctb committed:
update comment

ctb committed:
make lid_to_idx and idx_to_ident private

ctb committed:
moar comment

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
add sqlite clases to protocol tests

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
adjust protocol

ctb committed:
update to match protocol

ctb committed:
add, then hide, RevIndex test

ctb committed:
update the LCA_Database protocol

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/index_tests

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
SqliteCollectionManifest now passes all the tests

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
update row check to ignore _ prefixes

ctb committed:
implement remaining lca_db protocol for sqlite

ctb committed:
fix up rankinfo for sqlite LCA_Database

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/index_tests

ctb committed:
finish testing the rest of the Index classes

ctb committed:
cleanup

ctb committed:
upd

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
cleanup LCA_Database creation

ctb committed:
backport 08ac110dfad4afb76

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
add sqlite loading to CollectionManifest

ctb committed:
update manifest writing to support SQL, too

ctb committed:
switch to using generic manifest.write_to_filename

ctb committed:
catch pre-existing sqlite DBs

ctb committed:
remove test for now-implemented func

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
work through various merge implications

ctb committed:
switch away from a row tuple in CollectionManifest

ctb committed:
more clearly separate internals of LCA_Database from public API

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/index_tests

ctb committed:
add saved/loaded manifest

ctb committed:
add test coverage for exceptions in LazyLoadedIndex

ctb committed:
add docstrings to manifest code

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
add docstrings / comments

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
Merge branch 'add/sqlite_index' into add/sqlite_index_lca

ctb committed:
fix sig check reliance on internal manifest mechanism

ctb committed:
fix picklist stuff when using Sqlite manifests

ctb committed:
add lots of debug stmts

ctb committed:
remove SQLite pickset as impractical

ctb committed:
remove some expensive debugs

ctb committed:
remove sql picklist code as too slow

ctb committed:
comments and cleanup

ctb committed:
much cleanup

ctb committed:
re-add debug_literal

ctb committed:
more cleanup

ctb committed:
comment

ctb committed:
fix 'num' select

ctb committed:
test and document locations()

ctb committed:
use names in namedtuple; add containment test

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
add numerical values to jaccard order tests

ctb committed:
cleanup

ctb committed:
remove redundant tests

ctb committed:
test scaled=1 stuff pretty explicitly

ctb committed:
rename 'create_from_manifest' method

ctb committed:
cleanup

ctb committed:
add required_keys check

ctb committed:
add required_keys check

ctb committed:
check manifest equality only on required keys

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
add index tests for LCA_SqliteDatabase

ctb committed:
constructor/etc refactoring

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/index_tests

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
add scaled/dowsample test

ctb committed:
add downsample_scaled etc

ctb committed:
remove unused code

ctb committed:
cleanup

ctb committed:
Merge branch 'latest' into add/index_tests

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
update comment

ctb committed:
rename tables to have prefix sourmash_

ctb committed:
update with many a test

ctb committed:
fix diagnostic output during sourmash index #1949

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
handle bad versions of stuff

ctb committed:
update/simplify version checking

ctb committed:
add append test

ctb committed:
add notes about further tests

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/index_tests

ctb committed:
Merge branch 'add/index_tests' into add/sqlite_index

ctb committed:
minor comment update

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
fix after merge

ctb committed:
update table name for lineage db

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
more docs

ctb committed:
implement loading of LCA_SqliteDatabases at command line

ctb committed:
cleanup and testing

ctb committed:
start adding some documentation

ctb committed:
add location and manifest properties to LCA_SqliteDatabase

ctb committed:
update

ctb committed:
Merge farm:sourmash into add/sqlite_index

ctb committed:
update index protocol tests to check location, manifest

ctb committed:
add tests for fileinfo on all sql db variants

ctb committed:
add test for signatures_with_location

ctb committed:
upd

ctb committed:
add test of new-style lineage db file

ctb committed:
upd/cleanup

ctb committed:
try out inheritance instead of composition

ctb committed:
comment

ctb committed:
more cleanup

ctb committed:
clean up LCA_SqliteDatabase

ctb committed:
create some more tests...

ctb committed:
update checklist

ctb committed:
refactor and cleanup

ctb committed:
round out the tests a bit

ctb committed:
allow append

ctb committed:
cleanup, doc

ctb committed:
cleanup/simplify

ctb committed:
support picklists in LCA_Database.signatures

ctb committed:
fix up @CTB in LCA tests

ctb committed:
cleanup @CTB in test_cmd_signature

ctb committed:
add tests for picklist support in LCA_database.signatures()

ctb committed:
many minor updates

ctb committed:
more tests

ctb committed:
add more manifest tests

ctb committed:
add some final? tests

ctb committed:
one final test

ctb commented:
I think this is ready for review @sourmash-bio/devs! 🎉 

(What's another 3,000 lines of code between friends, anyway?)

mr-eyes commented on doc/command-line.md:
```diff
@@ -931,6 +931,15 @@ As of sourmash 4.2.0, `lca index` supports `--picklist`, to
+
+All sourmash commands work with either type of LCA database (the
+default JSON database, and the SQLite version). SQLite databases are
+larger than JSON databases on disk but are typically much faster and
+to load and search, and use much less memory.
```
```suggestion
larger than JSON databases on disk, but are typically much faster to load and search, and use much less memory.
```

ctb replied:
thanks! fixed in c05d044f8e44062

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
fix typo via @mr-eyes

ctb commented:
oops - I think I found a problem! `detect_types=PARSE_DECLTYPES` isn't needed! I'll verify later. (No need to hold review, removing it can only improve performance.)

ctb committed:
remove unnecessary PARSE_DECLTYPES

ctb committed:
Merge branch 'latest' into add/sqlite_index

mr-eyes committed:
Merge branch 'latest' into add/sqlite_index

mr-eyes commented:
I am pulling updates from `latest` to see how #1976 will affect the tests on gh-actions.

mr-eyes commented on src/sourmash/index/sqlite_index.py:
```diff
@@ -0,0 +1,1087 @@
+
+        sqlite_utils.add_sourmash_internal(cursor, 'SqliteManifest', '1.0')
+        cursor.execute("""
+        CREATE TABLE sourmash_sketches
```
```suggestion
        CREATE TABLE IF NOT EXISTS sourmash_sketches
```

ctb replied:
Yep, this was intentional; the append case is tested and handled elsewhere, so I wanted an error here!

mr-eyes commented on src/sourmash/tax/tax_utils.py:
```diff
@@ -831,7 +879,7 @@ class TEXT,
 
         # follow up and create index
-        cursor.execute("CREATE UNIQUE INDEX taxonomy_ident ON taxonomy(ident);")
+        cursor.execute("CREATE UNIQUE INDEX sourmash_taxonomy_ident ON sourmash_taxonomy(ident);")
```
```suggestion
        cursor.execute("CREATE UNIQUE INDEX IF NOT EXISTS sourmash_taxonomy_ident ON sourmash_taxonomy(ident);")
```

mr-eyes replied:
Running `sourmash lca index ....` twice with the same parameters will try to open and create tables and then threw exceptions. The failed table creation exception was handled in https://github.com/sourmash-bio/sourmash/blob/9e1ea60626a297c82c868e748ab7b0dcc16b8269/src/sourmash/tax/tax_utils.py#L879 which will prevent the process to complete. 

If `CREATE TABLE IF EXISTS` was added to that line, running `lca index` will open and update the same sqlite file and that will change the file content from the first run with now any thrown errors. 

**TL;DR** I think it would be better if `lca index` can detect that there's an already saved sqlite database and no need to re-run the command.

ctb replied:
I think for SQL databases, you are 100% right :). But `sourmash lca index` needs to handle JSON databases too, and there are also some issues around the logic for appending that would not be handled well in the JSON case and might require a lot of extra checks.

I ended up going a different route in c443ae10 -
* we now forbid overwriting / appending to existing LCA database;
* I added command-line and API tests for LCA database saving x 2 (all named `twice`);

mr-eyes commented:
Is building SBT SQLite index supported in the command line? 
`sourmash index` does not contain `-F {json,sql}, --database-format {json,sql}`.

mr-eyes commented on src/sourmash/lca/lca_db.py:
```diff
@@ -104,7 +119,11 @@ def _get_ident_index(self, ident, fail_on_duplicate=False):
 
     def _get_lineage_id(self, lineage):
-        "Get (create if nec) a unique lineage ID for each LineagePair tuples."
+        """Get (create if nec) a unique lineage ID for each
```
```suggestion
        """Get (create if necessary) a unique lineage ID for each
```

ctb commented:
> Is building SBT SQLite index supported in the command line? `sourmash index does not contain `-F {json,sql}, --database-format {json,sql}`.

`sourmash index` only builds SBTs; we've got an issue about changing that (https://github.com/sourmash-bio/sourmash/issues/949). 

SQLite databases (unlike SBT or LCA indexes) can be created with `sourmash sig cat ... -o xyz.sqldb`.

I agree this is confusing and inconsistent 😆 . I'll think about what to do - at the very least, better docs would be a good idea!

ctb committed:
Merge branch 'add/sqlite_index' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
add docs for creating sqldb

ctb commented:
> > Is building SBT SQLite index supported in the command line? `sourmash index does not contain `-F {json,sql}, --database-format {json,sql}`.

> I agree this is confusing and inconsistent 😆 . I'll think about what to do - at the very least, better docs would be a good idea!

updated docs in 16b8b9b7, and added some suggestions in #1949.

ctb committed:
do not allow overwrite/append to xisting lca database

ctb committed:
Update src/sourmash/lca/lca_db.py

Co-authored-by: Mohamed Abuelanin <mabuelanin@gmail.com>

ctb commented:
Thanks for catching the `lca index` problems with running twice!

mr-eyes commented:
There are inconsistent results when running `lca classify` on same lca database built with both `SQL` and `json`. You can reproduce here:

```bash
curl -L https://osf.io/bw8d7/download?version=1 -o delmont-subsample-sigs.tar.gz
tar xzf delmont-subsample-sigs.tar.gz
curl -O -L https://github.com/ctb/2017-sourmash-lca/raw/master/tara-delmont-SuppTable3.csv
sourmash lca index -f tara-delmont-SuppTable3.csv delmont.lca.json delmont-subsample-sigs/*.sig
sourmash lca index -F sql -f tara-delmont-SuppTable3.csv delmont.lca.sql delmont-subsample-sigs/*.sig
sourmash lca classify --db delmont.lca.json --query delmont-subsample-sigs/TARA_RED_MAG_00003.fa.gz.sig > json_results
sourmash lca classify --db delmont.lca.sql --query delmont-subsample-sigs/TARA_RED_MAG_00003.fa.gz.sig > sql_results
diff json_results sql_results
```

Using the SQLite database gave no matches, while there is a match using the JSON version.

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

ctb committed:
fix bug with duplicate lineages in LCA_SqliteDatabase

ctb commented:
> There are inconsistent results when running `lca classify` on same lca database built with both `SQL` and `json`.

WOW, very nice catch!

I gotta say I was pretty worried when I first saw this, but it turns out that rather than being some deep and structural flaw in how I was handling identifiers differently b/t SQL and JSON, it was just a straight up simple programming error 😅.

Fixed in 52bc90ba.

ctb committed:
fix test broken by duplicate lineage fix

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

mr-eyes commented:
> WOW, very nice catch!

Thank you! 

> I gotta say I was pretty worried when I first saw this.

Likewise! and was resisting myself to debugging xD Glad it was straightforward.

ctb commented:
> Likewise! and was resisting myself to debugging xD Glad it was straightforward.

it took me longer to track it down than I expected since I was gearing up for a big debugging effort for a subtle effect 😆

ctb committed:
Merge branch 'latest' of https://github.com/sourmash-bio/sourmash into add/sqlite_index

mr-eyes reviewed (approved):
This was a quite big PR to review, maybe I missed things or complicated-heuristical bugs while reviewing but I did a lot of testing and code inspection that I feel confident to merge.

I think it's good to wait for merging #2000 into this PR before merging to `latest`.

ctb commented:
> This was a quite big PR to review, maybe I missed things or complicated-heuristical bugs while reviewing but I did a lot of testing and code inspection that I feel confident to merge.

yay, thank you!

> I think it's good to wait for merging #2000 into this PR before merging to `latest`.

I'm going to merge it now, tho, since I haven't decided on the scope of #2000 yet; might be bigger than just a `convert` command!
